qiskit/__init__.py

```diff
@@ -12,42 +12,66 @@
 
 # pylint: disable=wrong-import-order,invalid-name,wrong-import-position
 
-
 """Main Qiskit public functionality."""
 
-import pkgutil
 import sys
 import warnings
 import os
 
 # First, check for required Python and API version
 from . import util
 
+# Extend namespace for backwards compat
+from qiskit import namespace
+new_meta_path = []
+new_meta_path.append(namespace.QiskitElementImport(
+    'qiskit_aer', 'qiskit.providers.aer'))
+new_meta_path.append(namespace.QiskitElementImport(
+    'qiskit_ignis', 'qiskit.ignis'))
+new_meta_path.append(namespace.QiskitElementImport(
+    'qiskit_aqua', 'qiskit.aqua'))
+new_meta_path.append(namespace.QiskitElementImport(
+    'qiskit_aqua.ai', 'qiskit.ai'))
+new_meta_path.append(namespace.QiskitElementImport(
+    'qiskit_aqua.chemistry', 'qiskit.chemistry'))
+new_meta_path.append(namespace.QiskitElementImport(
+    'qiskit_aqua.finance', 'qiskit.finance'))
+new_meta_path.append(namespace.QiskitElementImport(
+    'qiskit_aqua.ml', 'qiskit.ml'))
+new_meta_path.append(namespace.QiskitElementImport(
+    'qiskit_aqua.optimizations', 'qiskit.optimizations'))
+new_meta_path.append(namespace.QiskitElementImport(
+    'qiskit_ibmq_provider', 'qiskit.providers.ibmq'))
+new_meta_path.append(namespace.QiskitElementImport(
+    'qiskit_aqt_provider', 'qiskit.providers.aqt'))
+new_meta_path.append(namespace.QiskitElementImport(
+    'qiskit_honeywell_provider', 'qiskit.providers.honeywell'))
+# Add Qiskit importers to meta_path before PathFinder in the default
+# sys.meta_path to avoid the miss penalty on trying to import a module
+# which does not exist
+old_meta_path = sys.meta_path
+sys.meta_path = old_meta_path[:-1] + new_meta_path + [old_meta_path[-1]]
```

Comment:
I tested this a bit and didn't see a big difference in either order. Mostly because the new finders are no-ops if the namespace match isn't there. It adds the overhead of a single string comparison for each new finder which is minimal to loading from disk in practice. I used this order was there to try and avoid the lookup penalty for the old non-existent namespaced element paths searching the full list of python paths when we know that those are incorrect. But this really only matters for environments with painfully slow IO.